### PR TITLE
[CIS-204] Implement ChannelController

### PR DESCRIPTION
**Pull request by b-onc**
..actually "Update ChannelController" would be more suitable

The 2 controllers we have are so like each other, I think we should refactor `Controller` protocol!

### CODE REVIEW

File: StreamChatClient_v3/Controllers/ChannelController.swift [hunk 1]
```diff
@@ -5,21 +5,60 @@
...
-    private let viewContext: NSManagedObjectContext
-    private let worker: ChannelUpdater<ExtraData>
+    /// The `ChatClient` instance this controller belongs to.
+    public let client: Client<ExtraData>
+    
+    /// The channel matching the channelId. To observe updates to the channel, set your class as a delegate of this controller and call `startUpdating`.
+    public private(set) lazy var channel: ChannelModel<ExtraData>? = {
+        log.warning("Accessing `channels` before calling `startUpdating()` always results in nil channel.")
+        return nil
+    }()
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Controllers/ChannelController.swift:47]: I think the controller also needs to expose `[MessageModel]`. This `channel` variable is just to get the details about the channel itself.

File: StreamChatClient_v3/Controllers/ChannelController.swift [hunk 2]
```diff
@@ -5,21 +5,60 @@
...
+    
+    /// Used for observing the database for changes.
+    private(set) lazy var fetchResultsController: NSFetchedResultsController<ChannelDTO> = {
+        let request = ChannelDTO.fetchRequest(for: channelQuery.cid)
         let frc = NSFetchedResultsController(fetchRequest: request,
-                                             managedObjectContext: viewContext,
+                                             managedObjectContext: client.databaseContainer.viewContext,
                                              sectionNameKeyPath: nil,
                                              cacheName: nil)
         frc.delegate = self.changeAggregator
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Controllers/ChannelController.swift:64]: We will need multiple FRCs here. One for observing the changes in the channel itself, other for observing the list of messages.

File: StreamChatClient_v3/Controllers/ChannelController.swift [hunk 3]
```diff
@@ -28,55 +67,199 @@ public class ChannelController<ExtraData: ExtraDataTypes>: Controller {
...
         aggregator.onChange = { [unowned self] (_: [Change<ChannelModel<ExtraData>>]) in
             guard let channel = self.fetchResultsController.fetchedObjects?.first
                 .map(ChannelModel<ExtraData>.create(fromDTO:)) else { return }
-            
-            self.delegate?.channelController(self, didUpdateChannel: channel)
+            self.channel = channel
+            self.delegateCallback {
+                $0?.channelController(self, didUpdateChannel: channel)
+            }
         }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Controllers/ChannelController.swift:80]: The `ChangeAggregator` was designed for observing a list, so here, we'd ideally use it for observing the messages. We need to create something like `SingleEntityChangeAggregator`, which will provide a better API for observing a single entity.

File: StreamChatClient_v3/Controllers/ChannelController_Tests.swift [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+        client.databaseContainer.write { session in
+            try session.saveChannel(payload: self.dummyPayload(with: .unique), query: nil)
+        }
+        
+        // Assert the channel is not loaded
+        AssertAsync.staysTrue(controller.channel == nil)
+    }
+    
+    func test_startUpdating_fetchesExistingChannel() throws {
+        // Save two channels to DB (only one matching the query) and wait for completion
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Controllers/ChannelController_Tests.swift:69]: We don't use `query` in the tests at all. Probably some copy&paste leftover?

Comment (b-onc) [reply StreamChatClient_v3/Controllers/ChannelController_Tests.swift:69]: It's a copy/paste leftover but it's intentionally left. There _is_ a query, internally in controller. So the comment is correct: two channels are saved to db, but only one matches the ChannelController query

**Comment by VojtaStavik** [comment]
I'm gonna reopen a new PR with the udpated version soon

**Pull request closed without merging**